tests/Integration/Behaviour/Features/Context/Domain/CartRule/EditCartRuleFeatureContext.php

```diff
@@ -0,0 +1,354 @@
+            );
+        }
+    }
+
+    /**
+     * @param EditCartRuleCommand $command
+     * @param array<string, mixed> $data
+     */
+    private function fillCommand(EditCartRuleCommand $command, array $data): void
+    {
+        if (isset($data['name'])) {
+            $command->setLocalizedNames($data['name']);
+        }
+        if (isset($data['description'])) {
+            $command->setDescription($data['description']);
+        }
+        if (isset($data['highlight'])) {
+            $command->setHighlightInCart(PrimitiveUtils::castStringBooleanIntoBoolean($data['highlight']));
+        }
+        if (isset($data['allow_partial_use'])) {
+            $command->setAllowPartialUse(PrimitiveUtils::castStringBooleanIntoBoolean($data['allow_partial_use']));
+        }
+        if (isset($data['enabled'])) {
+            $command->setEnabled(PrimitiveUtils::castStringBooleanIntoBoolean($data['enabled']));
+        }
+        if (isset($data['code'])) {
+            $command->setCode($data['code']);
+        }
+        if (isset($data['customer'])) {
+            $command->setCustomerId(!empty($data['customer']) ? $this->getSharedStorage()->get($data['customer']) : 0);
+        }
+        if (isset($data['priority'])) {
+            $command->setPriority((int) $data['priority']);
+        }
+        if (isset($data['valid_from'])) {
+            $command->setDateRange(
+                new DateTimeImmutable($data['valid_from']),
+                new DateTimeImmutable($data['valid_to'])
+            );
+        }
+        if (isset($data['total_quantity'])) {
+            $command->setTotalQuantity((int) $data['total_quantity']);
+        }
+        if (isset($data['quantity_per_user'])) {
+            $command->setQuantityPerUser((int) $data['quantity_per_user']);
+        }
+        if (isset($data['minimum_amount'])) {
+            $command->setMinimumAmount(
+                $data['minimum_amount'],
+                $this->getSharedStorage()->get($data['minimum_amount_currency']),
+                PrimitiveUtils::castStringBooleanIntoBoolean($data['minimum_amount_tax_included']),
+                PrimitiveUtils::castStringBooleanIntoBoolean($data['minimum_amount_shipping_included'])
+            );
+        }
+
+        if (isset($data['discount_application_type'])) {
+            $command->setDiscountApplication(
+                $data['discount_application_type'],
+                // if specific product type is provided and product is not, then command should throw exception
+                isset($data['discount_product']) ? $this->getSharedStorage()->get($data['discount_product']) : null
+            );
+        }
+
+        if ($cartRuleAction = $this->buildAction($data)) {
+            $command->setCartRuleAction($cartRuleAction);
+        }
+    }
+
+    private function buildAction(array $data): ?CartRuleActionInterface
+    {
+        $actionWasSet = false;
+        $builder = new CartRuleActionBuilder();
```

Comment:
Ok for now but the `CartRuleActionBuilder` will have to be refactored to be come stateless later

Reply:
yeah Im still thinking about that stateless. Its hard to do. I've tried similar stuff, but its messy :smile: 
Further more if we just provide array of $data to build method, it will be coupled with form, meaning we will have to duplicate everything in behats :sweat: So I wonder if its really worth it

Reply:
Well you already have to do the mapping from the behat table format already, when you could also build the CartRuleAction from scratch without using the CartRuleActionBuilder

To make it work you would only need to adapt the behat format to the expected builder format:
```
$builderData = [
    'gift_product_id' => $tableNodeData['gift product'],
    'gift_product_attribute_id' => $tableNodeData['gift combination'],
];
```

It would actually simplify the behat test because only this mapping would be enough, you wouldn't need to replicate the behaviour of the form data handler: `I check that this field and this field is present, so I can call the builder setter with to attributes. Then I check that this fields and that is present and call another setter, ...`

So the whole building process would only be handled in the builder, of course its expected format would be based on the form structure because it's easier this way for us. But it's exactly what we did with the product builders, and even then we were able to adapt them to other use cases by relying on form data adapter classes (for combination bulk edition, or list edition)